news/_posts/2017-11-30-make-back-office-modules-great-again.md

```diff
@@ -0,0 +1,68 @@
+
+### Current situation of Back Office modules (pre 1.7.3)
+
+In Back Office, for the modern pages (migrated to Symfony 2), PrestaShop developers can still use Hooks like in the old ways. But they are not allowed to override templates, forms or data returned by the controllers.
+
+In Product page, a specific hook called `hookDisplayAdminProductsExtra` can display all modules into a specific tab and we have added some hooks at bottom, middle and top positions of each tab. However, it's not easy to hook at a specific position into the forms to add or re-order fields, buttons or blocks of information.
+
+Reviewing the actual hooks, they are not located wisely regarding the new UX of this page and some hooks are missing to make basic additions (like adding new buttons to menus for instance).
+
+![PrestaShop Hooks on Product Page](/assets/images/2017/11/product_page_hooks.png)
+
+Finally, and like in 1.6, it's still really difficult to know when, where and if a hook has been called, how many times and what are the modules hooked.
+
+### PrestaShop 1.7.3, the module experience re-invented
+
+The first new feature we'd like to present is the Hooks debugger. In all modern pages of Back Office, you have access to the PrestaShop debug toolbar, now enriched by a "Hooks" section.
+
+![PrestaShop Hooks Debugger section](/assets/images/2017/11/hooks_debugger_section.png)
+
+At first look, in every modern page you can know which and how many times hooks have been called. If you access the section panel, you have access to many additional information:
+
+* modules and hooks arguments
+* called and uncalled (but available) hooks
+* which modules react to hooks (the order is respected)
+
+![PrestaShop Hooks Debugger](/assets/images/2017/11/hooks_debugger.png)
+
+This allow every PrestaShop developer to choice wisely the right hook, but also to understand some behaviors if a specific hook doesn't render the expected view without have to enable/disable every module until find the guilty one.
+
+Then, we have introduced two hooks in Product Catalog page to make the contribution of menus simple:
+
+* `displayDashboardToolbarTopMenu` to be able to add new action to the top menu
+* `displayDashboardToolbarIcons` to allow the addition of new icons in Dashboard block (will be available in every Dashboard modern pages)
+
+![New Hooks](/assets/images/2017/11/new_hooks.png)
+
+One more think with PrestaShop 1.7.3, the introduction of **"modules as bundle" system**, allowing the override of templates, services (classes), commands, forms and everything from modern pages in your modules. We are documenting this new system right now, but regarding the templating override we can tell this will work mostly the same way we already work with override in Front Office theme templates.
```

Comment:
I'm not sure, I want to insist on an additional feature, "One more think" is fine, "à la Apple", but maybe I've misused it?

Reply:
"Another feature for PrestaShop 1.7.3/One more feature in PrestaShop 1.7.3: we have introduced a system called "modules as bundle" which lets the overrides of templates, services (classes), commands, forms, etc. used in the modern pages in your modules. We are still documenting this new system now but we can already tell you it will work the same way as it does with the front office theme templates."

Reply:
Sois mon scribe ! https://www.youtube.com/watch?v=qrYPyCPFRT0

Reply:
Ah ! Ah ! Quand tu veux : on dansera la vie, on chantera le code !